Add scripting

vieiraa opened the pull request:
Attempt to fix #22.
I only gonna work on this in my spare time, so this should take a while.

wheybags commented:
I'm not sure that now is the best time for this, and there are some issues to work out with whatever scripting language we use. The primary use case for it will be scripting in mods, and mods scripts will need to be deterministic just like the rest of the game code, in order for multiplayer to work. This means they will need to use our FixedPoint class for maths. Given that, I'm not sure if we shoudl try to hack FixedPoint into an existing scriping language, or if we should just make our own simplistic langauge. In particular, since FixedPoints can be a bit slow, it would be nice to use a language that distinguished between integer and real number types, so people could only use FixedPoint operations when they actually needed them.

wheybags commented:
That said, if you want to hack around and experiment with integrating FixedPoint into lua / other languages, please go ahead, but given the issues, I don't think it will be merged until we are looking adding proper mod support, probably after 1.0.

vieiraa commented:
I was planning to use this to implement the quest system. As for the FixedPoint class, it is fairly easy to integrate C++ classes into Lua, and Lua would execute C++ functions when we use the class' methods inside it.

wheybags commented:
The thing is, we would want lua to use that class for all of its maths. So, if a script tries to add two numbers, it should use FixedPoints internally instead of floats, which is not as simple as just adding a native binding for the class.

vieiraa commented:
Added FixedType registration for use in Lua. Also added unit tests to test it.

wheybags commented:
http://lua.2524044.n2.nabble.com/Fixed-point-LUA-NUMBER-td7630153.html is a good thread on the topic, and is talking more abotu what I'm getting at. We would need to bundle our own copy of lua (in the extern/ folder), and redefine it's internal number type using the LUA_NUMBER #define. The maths module will break, so we can just remove it for now. Or we could go for the other more moderate approach listed there of redefinign LUA_NUMBER to int32_t and keeping the custom fixedpoint class you've added.

vieiraa commented:
Lua 5.3 added support to integers, so I don't think it it necessary anymore.

wheybags commented:
It culd still be worth trying out defining LUA_NUMBER to FixedPoint, even in 5.3, see https://github.com/lua/lua/blob/v5.3.5/luaconf.h#L457, and https://github.com/lua/lua/blob/v5.3.5/luaconf.h#L127

This might be perfect actually, as the lua interpreter will try to use normal integer operations, but fall back to FixedPoint where necessary. Of course, we would also need to compile lua as c++, so that it could use the operator overloading of the FixedPoint class.

vieiraa commented:
Can I leave lua_Number as is, and just replace all operations involving floats (math functions, arithmetic operators) with FixedPoint ops? I think the result would be the same.

vieiraa commented:
Added FixedPoint to lua. Now all floating point operations and math functions use FixedPoint. Still needs to implement some missing math functions that are not present in FixedPoint (log, exp, ...), implement the modulo operator, and fix printing to stdout.

wheybags commented:
This is looking really great so far, thank you!
If all goes well, we might be able to start using this sooner that I had originally planned.

vieiraa commented:
Fixed printing the FixedPoint, and fixed a segfault when trying to stringify some numbers.
I have two questions:
1) Do we still need the class registration? (Being able to create c++ FixedPoint objects and access its methods)
2) How can I use FreeMono in the console? The Diablo font is too ugly for that, and there is no small caps.

vieiraa commented:
I'm having problems with the FixedPoint hashing in l_hashfloat. It causes a segfault in some cases. Would it be safe to hash the toDouble() value instead of the rawValue()?

vieiraa commented:
I'm having problems with the FixedPoint hashing in l_hashfloat, as it causes a segfault with some numbers. Any suggestions on how to solve this?

wheybags commented:
> Do we still need the FixedPoint class registration? (Being able to create c++ FixedPoint objects and access its methods from Lua)

I think we don't.

> How can I use FreeMono in the console? The Diablo font is too ugly for that, and there is no small caps.

You can set the font using the nuklear styling system. See here for an example of styling, I don't have an example of font setting specifically to hand.

wheybags commented:
https://github.com/Immediate-Mode-UI/Nuklear/blob/master/example/skinning.c#L420 Woops, forgot the link in my last post

vieiraa commented:
Ok, have added FixedPoint support to Lua, created a simple console and redirected stdout to it... what should I do next?

wheybags commented on extern/lua/lvm.cpp:
```diff
@@ -119,13 +119,16 @@ int luaV_tonumber_ (const TValue *obj, lua_Number *n) {
+      f += 1;               /* convert floor to ceil (remember: n != f) */
   }
-  return lua_numbertointeger(f, p);
+  *p = f.intPart();
```
Why not leave in the call to lua_numbertointeger here?

wheybags commented on extern/lua/lvm.cpp:
```diff
@@ -652,7 +656,7 @@ void luaV_concat (lua_State *L, int total) {
       /* collect total length and number of strings */
       for (n = 1; n < total && tostring(L, s2v(top - n - 1)); n++) {
-        size_t l = vslen(s2v(top - n - 1));
+        size_t l = vslen(s2v(top - n));
```
Is this intentional?

wheybags commented on extern/lua/lvm.cpp:
```diff
@@ -1133,29 +1135,29 @@ void luaV_execute (lua_State *L, CallInfo *ci) {
 #if LUA_USE_JUMPTABLE
 #include "ljumptab.h"
 #endif
- tailcall:
```
A lot of formatting only changes here

vieiraa commented on apps/freeablo/fagui/console/console.cpp:
```diff
@@ -0,0 +1,100 @@
+        bufferLen += inputLen;
+        inputLen = 0;
+
+        if (Misc::StringUtils::ciStartsWith(command, "eval"))
```
Eval is just to evaluate pure lua code. Game commands would be fetch from resources/commands.

wheybags replied:
I know, I'm saying I don't think we want that command fetching logic

vieiraa commented on extern/lua/luaconf.h:
```diff
@@ -433,59 +434,9 @@
-
+#define LUA_NUMBER FixedPoint
+#define LUAI_UACNUMBER FixedPoint
+#define l_floatatt(n) (DBL_##n)
```
Right now it is only being used on the random generators in lmathlib.cpp. Can I use C++ random instead? I don't think we have to update this file, since it was heavily modified.

wheybags replied:
We can just remove them for now, if we need them later we can add them then

wheybags replied:
Using our Random component would be fine too though

vieiraa commented on extern/lua/lobject.cpp:
```diff
@@ -220,13 +220,15 @@ static lua_Number lua_strx2number (const char *s, char **endptr) {
+  char* endptr;
+  *result = lua_Number(s);
+  if (endptr == s)
+    return NULL; /* nothing recognized? */
```
Can I replace the exception with std::optional? This way we can check if the value is "nil" in Lua, if the conversion fails.

wheybags replied:
No, we can't remove the exceptions. They are in a constructor, so there is no way to return an std::optional. We could maybe add a FixedPoint::tryParseFromString() or something similar that could return an optional though. Internally I guess it would just catch the exception. I believe it should already be working, but since we're now relying on it, we should also add a test to make sure FixedPoint doesn't accept trailing characters. Thins function does expect it to accept trailing whitespace though, so that should be handled inside this function as well (I wouldn't add support for that to FixedPoint directly).
Also, we need to return the "endptr". What you are doing atm is undefined behaviour, because you're returning the c_str() from an std::string which will be destroyed when exiting the function. It seems the code that calls this function is only ever checking it against NULL, so maybe an easy fix would be to change the return type to a bool indicating success instead.

vieiraa commented on apps/freeablo/fagui/console/console.cpp:
```diff
@@ -0,0 +1,100 @@
+    {
+        std::cout.rdbuf(consoleCout.rdbuf());
+        freopen(nullFile, "a", stdout);
+        setbuf(stdout, consoleStdoutBuffer);
```
I'm gonna leave this to a future PR.

vieiraa commented on extern/lua/lstrlib.cpp:
```diff
@@ -1072,7 +1072,7 @@ static int lua_number2strx (lua_State *L, char *buff, int sz,
 ** is maximum exponent + 1). (99+3+1, adding some extra, 110)
 */
-#define MAX_ITEMF	(110 + l_floatatt(MAX_10_EXP))
+#define MAX_ITEMF	(110)// + l_floatatt(MAX_10_EXP))
```
This MAX_10_EXP macro is the float type maximum exponent (37 or greater). What would FixedPoint's be?

wheybags replied:
I had to think about this one a bit :p
It should be:
```
ceil(log10(INT64_MAX / FixedPoint::scalingFactor)) + FixedPoint::scalingFactorPowerOf10 + 2
^ Before decimal                                     ^ After decimal                      ^ - and .
```

We can probably just precalculate this value (which is 21), and put it in FixedPoint::maxStringLength as a constexpr variable beside scalingFactor and scalingFactorPowerof10, with a comment explaining where the value came from.

vieiraa replied:
Should I use FixedPoint's ceil and log, or cmath's?

vieiraa commented:
I'm getting "ERROR INVALID SPRITE CACHE REQUEST 1" when I try to open the console since #429 got merged.

vieiraa commented:
Wtf? PR just closed on its own.

khanduras commented:
Got me all excited, it'll be fun to play with this once it's ready for an audience

wheybags commented:
Just so you know, I think things are looking decent here, and I haven't forgotten about the PR, but for now it won't be merged. There will come a time in the implementation of the game when we will need a scripting engine, and at that point this will be pulled in. I'm not sure when that will be, maybe when we first start to add mods, or quests or some other point, but it will happen. For now though, there's no need to have it in the main tree before it's in use.
Thanks again for all the work!

khanduras commented:
Is scripting done? Is it working? I'm assuming it's using LUA?

wheybags commented:
This is a modified version of lua which is mostly ready to be integrated. Scripting being "donw" would mean that ypu can actually do something with it (like add custom lua scripts into the game), which is not yet done, and won't be for a while, as the priority is getting the base game working properly.
